aws/resource_aws_codepipeline_webhook.go

```diff
@@ -0,0 +1,336 @@
+	var conf codepipeline.WebhookAuthConfiguration
+	switch authType {
+	case codepipeline.WebhookAuthenticationTypeIp:
+		ipRange := authConfig["allowed_ip_range"].(string)
+		if ipRange == "" {
+			return nil, fmt.Errorf("An IP range must be set when using IP-based auth")
+		}
+
+		conf.AllowedIPRange = &ipRange
+
+		break
+	case codepipeline.WebhookAuthenticationTypeGithubHmac:
+		secretToken := authConfig["secret_token"].(string)
+		if secretToken == "" {
+			return nil, fmt.Errorf("Secret token must be set when using GITHUB_HMAC")
+		}
+
+		conf.SecretToken = &secretToken
+		break
+	case codepipeline.WebhookAuthenticationTypeUnauthenticated:
+		break
+	default:
+		return nil, fmt.Errorf("Invalid authentication type %s", authType)
+	}
+
+	return &conf, nil
+}
+
+func resourceAwsCodePipelineWebhookCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).codepipelineconn
+	authType := d.Get("authentication").(string)
+
+	authConfig, err := extractCodePipelineWebhookAttr(d, "authentication_configuration")
+	if err != nil {
+		return err
+	}
+
+	rules, err := extractCodePipelineWebhookRules(d.Get("filter").(*schema.Set))
+	if err != nil {
+		return err
+	}
+
+	conf, err := extractCodePipelineWebhookAuthConfig(authType, authConfig)
+	if err != nil {
+		return err
+	}
+
+	request := &codepipeline.PutWebhookInput{
+		Webhook: &codepipeline.WebhookDefinition{
+			Authentication:              aws.String(authType),
+			Filters:                     rules,
+			Name:                        aws.String(d.Get("name").(string)),
+			TargetAction:                aws.String(d.Get("target_action").(string)),
+			TargetPipeline:              aws.String(d.Get("target_pipeline").(string)),
+			AuthenticationConfiguration: conf,
+		},
+	}
+
+	webhook, err := conn.PutWebhook(request)
+	if err != nil {
+		return fmt.Errorf("Error creating webhook: %s", err)
+	}
+
+	arn := *webhook.Webhook.Arn
+	d.SetId(arn)
+
+	url := *webhook.Webhook.Url
+	d.Set("url", url)
+
+	return resourceAwsCodePipelineWebhookRead(d, meta)
+}
+
+func getAllCodePipelineWebhooks(conn *codepipeline.CodePipeline) ([]*codepipeline.ListWebhookItem, error) {
+	var webhooks []*codepipeline.ListWebhookItem
+	var nextToken string
+
+	for {
+		input := &codepipeline.ListWebhooksInput{
+			MaxResults: aws.Int64(int64(60)),
+		}
+		if nextToken != "" {
+			input.NextToken = aws.String(nextToken)
+		}
+
+		out, err := conn.ListWebhooks(input)
+		if err != nil {
+			return webhooks, err
+		}
+
+		webhooks = append(webhooks, out.Webhooks...)
+
+		if out.NextToken == nil {
+			break
+		}
+
+		nextToken = aws.StringValue(out.NextToken)
+	}
+
+	return webhooks, nil
+}
+
+func setCodePipelineWebhookFilters(webhook codepipeline.WebhookDefinition, d *schema.ResourceData) error {
+	filters := []interface{}{}
+	for _, filter := range webhook.Filters {
+		f := map[string]interface{}{
+			"json_path":    *filter.JsonPath,
```

Comment:
To prevent potential panics, all dereferences should be handled by the AWS Go SDK provided helper functions or have `nil` checks. e.g. `"json_path": aws.StringValue(filter.JsonPath),`

This is true of many of the `*` dereferences below, but I'll withhold from commenting on every single one of them.